plugin-server/src/main/ingestion-queues/analytics-events-ingestion-consumer.ts

```diff
@@ -46,3 +68,61 @@ export const startAnalyticsEventsIngestionConsumer = async ({
 
     return queue
 }
+
+export async function eachBatchIngestionWithOverflow(
+    payload: EachBatchPayload,
+    queue: IngestionConsumer
+): Promise<void> {
+    function groupIntoBatchesIngestion(kafkaMessages: KafkaMessage[], batchSize: number): KafkaMessage[][] {
+        // Once we see a distinct ID we've already seen break up the batch
+        const batches = []
+        const seenIds: Set<string> = new Set()
+        let currentBatch: KafkaMessage[] = []
+        for (const message of kafkaMessages) {
+            const pluginEvent = formPipelineEvent(message)
+            const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
+            if (currentBatch.length === batchSize || seenIds.has(seenKey)) {
+                seenIds.clear()
+                batches.push(currentBatch)
+                currentBatch = []
+            }
+            seenIds.add(seenKey)
+            currentBatch.push(message)
```

Comment:
Feel free to ignore (& if so maybe add a comment): It would be better if we didn't split by the ids going to overflow, they can all be processed in parallel, which would be faster and help us recover faster if they take up the majority of a partition. So ideally we'd lookup here if this one is going to overflow, if yes then mark it somehow (so we don't need to lookup twice - e.g. maybe set the key to be null) & don't add to the `seenKey` set (we don't want this to break up the batch). Though there is a problem here if we didn't consume the message in the end for various reasons it still gets added to the limiter.

```suggestion
            const pluginEvent = formPipelineEvent(message)
            if (message.key == null || ConfiguredLimiter.consume(message.key.toString(), 1) === false) {
                // set message key to be null so we know to send it to overflow topic, we don't want to do it here to preserve the kafka offset handling
                message.key = null
                seenKey = null
            } else {
                const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
            }
            if (currentBatch.length === batchSize || (seenKey != null && seenIds.has(seenKey))) {
                seenIds.clear()
                batches.push(currentBatch)
                currentBatch = []
            }
            seenIds.add(seenKey)
            currentBatch.push(message)
```